### PR TITLE
vcat of NullableCategoricalArrays failed with nulls (refs to 0)

### PR DESCRIPTION
**Pull request by gustafsson**
Didn't have a test for this before.

### CODE REVIEW

File: src/array.jl [hunk 1]
```diff
@@ -211,7 +211,7 @@ end
                 r
             end for a in A]
             $A(DefaultRefType[refs...;],
-               CategoricalPool(newlevels, ordered && all(isordered, A)))
+               CategoricalPool(newlevels, ordered && all(a->isordered(a) || 0==length(levels(a)), A)))
```
Comment (nalimilan) [inline src/array.jl]: `isempty(levels(a)) || ...` should be slightly more efficient (and more concise).

File: test/12_nullablearray.jl [hunk 2]
```diff
@@ -970,8 +970,18 @@ for ordered in (false, true)
...
            NullableCategoricalArray(["a","","","a"],[false,true,true,false]))
 
+# vcat with all nulls preserves isordered
+ca1 = NullableCategoricalArray(["a","b"],[false,true])
+ordered!(ca1, true)
+ca2 = NullableCategoricalArray(["a","b"],[true,true])
+ordered!(ca2, false)
+r = vcat(ca1,ca2)
+@test isequal(r, NullableCategoricalArray(["a","","",""],[false,true,true,true]))
+@test isordered(r)
```
Comment (nalimilan) [inline test/12_nullablearray.jl]: Wait, concatenating an ordered and an unordered array gives an ordered array? Should rather be the contrary. Though changing this can wait for another PR.

**Comment by nalimilan** [comment]
Could be useful to add tests for uncovered cases like:
- `CategoricalArray` with `#undef` values
- arrays with only null/undef values
- empty array(s) with no levels
- empty array(s) with levels
- matrices

**Comment by nalimilan** [comment]
Any chance you could update the PR? I won't tag a new release until this regression is fixed.

File: test/12_nullablearray.jl (hunk 2 above)
Comment (gustafsson) [inline test/12_nullablearray.jl]: The ideas is that concatenation of an array with only nulls shouldn't affect the ordering. But I'll postpone that to another PR.

**Comment by gustafsson** [comment]
* `CategoricalArray` or `NullableCategoricalArray` with `#undef` doesn't work, but it's not related to this PR. I tested with: `CategoricalArray(Array{String}(1))` and `NullableCategoricalArray(Array{String}(1),[true])`
* Matrices are covered in [13_arraycommon.jl#L38](https://github.com/nalimilan/CategoricalArrays.jl/blob/master/test/13_arraycommon.jl#L38)
* The other tests you suggested are covered by the added commits.

File: src/array.jl [hunk 3]
```diff
@@ -298,6 +298,20 @@ end
...
         convert(::Type{$A}, A::$A) = A
+
+
+        ## Other methods
+        function vcat(A::$A...)
+            newlevels, ordered = mergelevels(map(levels, A)...)
+
+            refs = [begin
+                r = a.refs
+                r[r .> 0] = indexin(index(a.pool), newlevels)[r[r .> 0]]
```
Comment (nalimilan) [inline src/array.jl]: I don't find multiline comprehensions very easy to read. Also, `r .> 0` is going to allocate a lot: better unroll this operation. So you should allocate `refs` beforehand, and then loop over it.

File: test/12_nullablearray.jl [hunk 4]
```diff
@@ -970,4 +970,27 @@ for ordered in (false, true)
     end
 end
 
+# Test vcat with nulls
+@test isequal(vcat(NullableCategoricalArray(["a","b"],[false,true]),
+           NullableCategoricalArray(["b","a"],[true,false])),
```
Comment (nalimilan) [inline test/12_nullablearray.jl]: You can use `==` instead of  `isequal` since I overwrote it at the top of the file for simplicity.

File: test/12_nullablearray.jl [hunk 5]
```diff
@@ -970,4 +970,27 @@ for ordered in (false, true)
     end
 end
 
+# Test vcat with nulls
```
Comment (nalimilan) [inline test/12_nullablearray.jl]: Please test `levels` and `isordered` on the result. That's really an essential property of categorical arrays.

File: test/12_nullablearray.jl (hunk 2 above)
Comment (nalimilan) [inline test/12_nullablearray.jl]: Ah, OK. That's an interesting corner-case indeed. I still think an unordered array should give unordered results even when empty, but we can discuss that elsewhere.

**Comment by nalimilan** [comment]
Thanks. I just noticed you forgot to remove the old `vcat` function, though.

**Comment by nalimilan** [comment]
Thanks!
